core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala

```diff
@@ -661,35 +667,51 @@ private[spark] class TaskSchedulerImpl(
         }
 
         if (launchedAnyTask && taskSet.isBarrier) {
+          val barrierPendingLaunchTasks = taskSet.barrierPendingLaunchTasks.values.toArray
           // Check whether the barrier tasks are partially launched.
-          // TODO SPARK-24818 handle the assert failure case (that can happen when some locality
-          // requirements are not fulfilled, and we should revert the launched tasks).
-          if (addressesWithDescs.size != taskSet.numTasks) {
-            val errorMsg =
-              s"Fail resource offers for barrier stage ${taskSet.stageId} because only " +
-                s"${addressesWithDescs.size} out of a total number of ${taskSet.numTasks}" +
-                s" tasks got resource offers. This happens because barrier execution currently " +
-                s"does not work gracefully with delay scheduling. We highly recommend you to " +
-                s"disable delay scheduling by setting spark.locality.wait=0 as a workaround if " +
-                s"you see this error frequently."
-            logWarning(errorMsg)
-            taskSet.abort(errorMsg)
-            throw new SparkException(errorMsg)
-          }
+          if (barrierPendingLaunchTasks.size != taskSet.numTasks) {
+            barrierPendingLaunchTasks.foreach { task =>
```

Comment:
nit: `barrierPendingLaunchTasks.reverse`

Reply:
hmm...I may miss something here. but why is `reverse` better? The `barrierPendingLaunchTasks` is a `HashMap` and its `values` doesn't have a meaningful order..so why we need `reverse` here?

Reply:
Good point - did not pay attention that it was a map.
My concern is as follows, reverting the availableResources (by prepending to the begining) ends up changing the relative ordering within this loop.

Thoughts ?

Reply:
I see, I got your point.

By `prependAll`, I was trying to restore the order too but it's still wrong. Probably, sorting tasks by taskId before reverting should restore the right order.

But after a second thought, the order actually doesn't matter during scheduling. Do you think it's still better for us to restore the order?

Reply:
Whether the order matters is a good question - I am not sure actually ... given the care taken in earlier code to restore state, I was assuming it might matter ?
I will let @tgravescs and @jiangxb1987 to comment more on it.
If the order does not matter, then I agree the existing change is fine - but then prependAll is also not required (append is fine).

Reply:
IIUC, the order shouldn't matter. Changed to `appendAll` first. Let's see @tgravescs @jiangxb1987 's opinion.

Reply:
the order of the addresses at this point doesn't matter. It does matter inside of resourceOfferSingleTaskSet

Reply:
Thx for confirming Tom !